File: app/api/browsers.py
from fastapi import APIRouter, HTTPException

from app.core.bitbrowser import BitBrowserClient, BitBrowserError
from app.core.browser_sessions import BrowserSessionManager
from app.schemas import BrowserEnsureRequest, BrowserOpenResponse

router = APIRouter(prefix="/api/browsers", tags=["browsers"])


@router.get("")
def list_browsers():
    try:
        return BitBrowserClient().list_profiles()
    except BitBrowserError as exc:
        raise HTTPException(status_code=503, detail=str(exc)) from exc


@router.post("/ensure")
def ensure_browser(data: BrowserEnsureRequest):
    try:
        return BrowserSessionManager().ensure_browser_for_account(
            data.account_id,
            template_browser_id=data.template_browser_id,
            rotate_proxy=data.rotate_proxy,
        )
    except Exception as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.post("/{browser_id}/open", response_model=BrowserOpenResponse)
def open_browser(browser_id: str):
    try:
        data = BrowserSessionManager().open_browser(browser_id)
        return {"browser_id": browser_id, "ws": data.get("ws")}
    except Exception as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.post("/{browser_id}/close")
def close_browser(browser_id: str):
    try:
        BrowserSessionManager().close_browser(browser_id)
        return {"message": "closed"}
    except Exception as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

File: frontend/src/api/browsers.ts
import { jsonBody, request } from "./client";
import type { BrowserEnsureRequest, BrowserOpenResponse } from "./types";

export function listBrowsers() {
  return request<unknown>("/api/browsers");
}

export function ensureBrowser(payload: BrowserEnsureRequest) {
  return request<{ browser_id: string; profile?: unknown; proxy_url?: string }>("/api/browsers/ensure", { method: "POST", body: jsonBody(payload) });
}

export function openBrowser(browserId: string) {
  return request<BrowserOpenResponse>(`/api/browsers/${browserId}/open`, { method: "POST" });
}

export function closeBrowser(browserId: string) {
  return request<{ message: string }>(`/api/browsers/${browserId}/close`, { method: "POST" });
}
